test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -79,167 +82,313 @@ func (s *snapshottableTestSuite) SkipRedundantSuite(driver TestDriver, pattern t
 }
 
 func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	f := framework.NewDefaultFramework("snapshotting")
+
+	type local struct {
+		config        *PerTestConfig
+		driverCleanup func()
+		cleanupSteps  []func()
+
+		cs        clientset.Interface
+		dc        dynamic.Interface
+		pvc       *v1.PersistentVolumeClaim
+		sc        *storagev1.StorageClass
+		pod       *v1.Pod
+		vsc       *unstructured.Unstructured
+		vs        *unstructured.Unstructured
+		vscontent *unstructured.Unstructured
+	}
 	var (
 		sDriver SnapshottableTestDriver
 		dDriver DynamicPVTestDriver
+		l       local
 	)
 
-	ginkgo.BeforeEach(func() {
-		// Check preconditions.
-		framework.ExpectEqual(pattern.SnapshotType, testpatterns.DynamicCreatedSnapshot)
-		dInfo := driver.GetDriverInfo()
-		ok := false
-		sDriver, ok = driver.(SnapshottableTestDriver)
-		if !dInfo.Capabilities[CapSnapshotDataSource] || !ok {
-			e2eskipper.Skipf("Driver %q does not support snapshots - skipping", dInfo.Name)
-		}
-		dDriver, ok = driver.(DynamicPVTestDriver)
-		if !ok {
-			e2eskipper.Skipf("Driver %q does not support dynamic provisioning - skipping", driver.GetDriverInfo().Name)
+	// SnapshotClassTest represents parameters to be used by snapshot tests.
+	// Not all parameters are used by all tests.
+	type SnapshotClassTest struct {
+		DeletionPolicy string
+	}
+
+	// TestSnapshottable tests volume snapshots based on a given SnapshotClassTest
+	TestSnapshottable := func(sct SnapshotClassTest) {
```

Comment:
Is there a reason why we need to declare these as closures instead of a normal function? I find that it's easier to read and reason about things like variable scoping as a normal function.

Reply:
I had it as a closure because `local` was a struct local to `TestSnapshottable`. Reworked this to be a standalone function.